text/0000-extended-standard-library.md

```diff
@@ -0,0 +1,370 @@
+
+# Appendix A
+[appendix-a]: #appendix-a
+
+## Example Policies
+
+A simplified list of example policies to use as potential starting point.
+
+### Evolving APIs / Breaking Changes
+
+ESL crates will be versioned with [Semantic Versioning](https://semver.org/).
+
+Patch and minor updates may happen at any time.
+
+Major updates will happen no more than once per year unless necessitated by a significant security concern.
+
+### General Support Policy
+
+Each major release of an ESL crate will receive backwards-compatible updates, at least addressing security and stability issues, for two years after the release of the subsequent major release. For example - if version 2.x.x is released on January 1, 2027, then version 1.x.x will receive updates until January 1, 2029.
+
+The only exception to this policy is if resolving a security issue requires a new major release (i.e., cannot be fixed in a backwards-compatible manner). If this happens, all previous vulnerable major versions will cease to receive updates immediately.
+
+### Security Updates
+
+The ESL security policy should largely mirror the [Go Security Policy](https://go.dev/doc/security/policy).
+
+Reports will be acknowledged within 7 days, and each vulnerability will be assigned a track - PUBLIC, PRIVATE, or URGENT. All security issues will be issued CVE numbers.
+
+Security updates will be created and released as quickly as reasonably possible.
+
+URGENT track issues are fixed in private and trigger an immediate dedicated security release, possibly with no pre-announcement.
+
+### Code Review and Testing Standards
+
+Rust ESL code will be managed in GitHub repositories. GitHub will be used for managing issues and pull requests.
+
+The Rust ESL project will maintain a list of maintainers who are qualified to review PRs.
+
+Any change to ESL code must be reviewed by a maintainer who is not the patch author. This will be enforced by GitHub settings. Pull requests will be required, no direct merges will be allowed.
+
+Whenever possible, changes to functionality should come with tests. When it is reasonably possible to write a test, no changes to functionality will be merged until they are merged with tests.
+
+Rust code in the ESL will follow the [Rust Style Guide](https://github.com/rust-lang/rust/tree/HEAD/src/doc/style-guide/src).
+
+API styles for ESL crates should become consistent over time. At minimum, it should be a goal to converge on the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/about.html). This does not need to happen prior to a crate joining the ESL, but ESL APIs should evolve towards this ideal at an acceptable rate of change.
+
+### Minimum Supported Rust Version (MSRV) Guarantees
+
+A maximum MSRV will be selected for the entire ESL and updated as it makes sense to do so. The maximum MSRV will never be newer than two years old. Individual crates may have MSRVs older than the maximum.
```

Comment:
Not to debate policy but just so I don't forget, consider the MSRV applying to the package and not every version.

What I mean is that a previous but supported minor satisfying the MSRV should count, even if the latest version doesn't. Maintainers will have flexibility to bump MSRV while having an incentive to not do it too often because it means more versions to support, creating more work, see https://github.com/rust-lang/rfcs/pull/3810/files#r2080681065

Reply:
Similarly about not wanting to forget, consider whether "N versions backs" actually meets people's needs. That means the MSRV can change every 6 weeks. Those with a direct MSRV requirement usually need predictable, stable stuff. Calendar based approaches might work better.

Reply:
If were legislating MSRV we'd need to actually pick a policy on cargo features affecting MSRV, which means the cargo resolver should support that logic as well. Not that we can't do any of that, but it's complicated

Reply:
From where I am currently out, I see per-feature MSRVs to be highly unlikely.

Reply:
If that can't be supported in some way, then "two years of msrv" is effectively a joke proposal.

Reply:
How much would `cfg(version)` and `cfg(accessible)` reduce the need for it?

Reply:
How much would `cfg(version)` and `cfg(accessible)` reduce the need for it?

Reply:
Some? But like imagine `min_const_generics` goes Stable and then the *entire* ESL crate group doesn't touch it at all for 2 years, not even opt-in. That's not a serious proposal. Who would possibly accept that?

Reply:
There are projects in the ecosystem already doing something like this. Many projects appear to (either as policy or as de facto policy) track Debian stable. Which isn't much different than having a 2 year old MSRV.

Reply:
To give a more concrete example: `bytemuck` with only default features builds as far back as 1.34, but various features you can enable add any number of things that didn't exist way back then. There's even a `latest_stable_rust` feature to enable like the dozen other features that have accumulated over the years.

It's completely unrealistic to construct official policy that says (1) we'll support a given MSRV for 2+ years and also (2) cargo features should never affect the MSRV of the crate.

Reply:
Oh I see. I don't see much value in picking apart this specific point. I think this is just an example policy? Like I probably wouldn't sign off on this either personally, but I'd pick that battle once we actually get to the point of setting policies like this. And in that context, yeah I would for sure argue in favor of being able to have different MSRVs for different feature combos.

Reply:
epage was just taking notes on what the ESL org would have to cover eventually, I'm mostly just adding to the notes pile.

and i know that RFCs aren't always implemented exactly as first written, but if there's not really a remotely realistic way to implement what the RFC says, then the RFC needs to say something else. The current RFC text currently states a pretty specific policy.

Reply:
Okay.

Reply:
Along the lines of MSRVs is the conflict between long MSRVs and build times which is also another strong care about.  I was analyzing someones dependencies recently and there were a lot of deps that exist solely for MSRV.  Some could be handled through polyfills while others will require `cfg(version)` / `cfg(accessible)` to be stabilized.  This is another reason to consider MSRV to be about the project as a whole and not apply to every major or minor version.

Reply:
`no_std` support is another major source of otherwise unnecessary dependencies. Many projects unconditionally depend on libraries like `libm` and `hashbrown` because there is no ergonomic way to specify dependencies that are only enabled when `not(feature = "std")`. So you either have to add something like a `no_std` or `libm` feature (which is inconvenient for consumers of your crate) or unconditionally depend on the necessary crates.

Reply:
I wonder what all cases there are like this and if there are some high leverage changes that could be made to avoid this, like getting `HashMap` into `alloc` with a no-std hasher available.